File: tests/oracle/block_index_assign_1204.cc
/*
 * BLOCK-INDEX-ASSIGN-0001: locked Ghidra 12.0.4
 * BlockGraph::findSpanningTree (block.cc:1009-1136) reverse-post-order
 * FlowBlock::index assignment oracle.
 *
 * The fixture builds synthetic control-flow graphs through the production
 * BlockGraph APIs (newBlock/addEdge) and observes the complete post-call
 * state: preorder and rootlist output vectors, the reordered component
 * list, every block's index / visitcount / numdesc / copymap, every
 * out-edge and mirrored in-edge label, and the distinct-in-range index
 * invariant.
 *
 * Cases cover: empty graph early return, single block, single-entry DAG
 * with pre-corrupted index/visitcount/copymap/numdesc, multi-entry
 * rootlist head/tail swap, no-root assume-first with a loop, an
 * unreachable component forcing the two-pass extraroots regeneration, the
 * stale-root list machinery with self-loops, and irreducible/goto
 * pre-labels that the pass-start clearEdgeFlags(~0) wipes.
 */

#include <bits/stdc++.h>

// Test-only access: FlowBlock's private fields sit behind an explicit
// `private:` label, but BlockGraph's helper members (findSpanningTree,
// clearEdgeFlags, ...) are implicitly private (no access label after
// `class BlockGraph : public FlowBlock {`), so `private -> public` alone
// cannot reach them.  `class -> struct` opens the implicit-private section;
// no decompile header uses `template<class ...>` or `enum class`, and
// <bits/stdc++.h> has already pulled every standard header, so the rewrite
// only touches Ghidra declarations in this translation unit.
#define private public
#define class struct
#include "architecture.hh"
#include "block.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef class
#undef private

using namespace ghidra;

static std::string flagsText(uint4 label)
{
  std::string out;
  if (label & FlowBlock::f_goto_edge) out += 'g';
  if (label & FlowBlock::f_loop_edge) out += 'l';
  if (label & FlowBlock::f_defaultswitch_edge) out += 'd';
  if (label & FlowBlock::f_irreducible) out += 'i';
  if (label & FlowBlock::f_tree_edge) out += 't';
  if (label & FlowBlock::f_forward_edge) out += 'f';
  if (label & FlowBlock::f_cross_edge) out += 'c';
  if (label & FlowBlock::f_back_edge) out += 'b';
  if (label & FlowBlock::f_loop_exit_edge) out += 'x';
  if (out.empty()) out.push_back('-');
  return out;
}

class Graph {
public:
  BlockGraph graph;
  std::vector<FlowBlock *> created;

  FlowBlock *makeBlock(void)
  {
    FlowBlock *bl = graph.newBlock();
    created.push_back(bl);
    return bl;
  }

  void edge(FlowBlock *from, FlowBlock *to)
  {
    graph.addEdge(from, to);
  }

  std::string name(FlowBlock *bl) const
  {
    std::ostringstream out;
    out << 'b' << (std::find(created.begin(), created.end(), bl) - created.begin());
    return out.str();
  }

  void observe(const std::string &caseName,
               std::vector<FlowBlock *> &preorder,
               std::vector<FlowBlock *> &rootlist)
  {
    std::ostringstream pre, roots, list, blocks, outedges, inedges;
    pre << '[';
    for (int4 i = 0; i < preorder.size(); ++i) {
      if (i != 0) pre << ',';
      pre << name(preorder[i]);
    }
    pre << ']';
    roots << '[';
    for (int4 i = 0; i < rootlist.size(); ++i) {
      if (i != 0) roots << ',';
      roots << name(rootlist[i]);
    }
    roots << ']';
    const std::vector<FlowBlock *> &gl = graph.getList();
    list << '[';
    for (int4 i = 0; i < gl.size(); ++i) {
      if (i != 0) list << ',';
      list << name(gl[i]);
    }
    list << ']';
    bool indices_ok = true;
    std::vector<bool> seen(created.size(), false);
    blocks << '[';
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      if (i != 0) blocks << ';';
      blocks << name(bl) << ":index=" << bl->index
             << ",visit=" << bl->visitcount
             << ",desc=" << bl->numdesc
             << ",copy=";
      if (bl->copymap == bl)
        blocks << "self";
      else if (bl->copymap == (FlowBlock *)0)
        blocks << "null";
      else
        blocks << "other:" << name(bl->copymap);
      if (bl->index < 0 || bl->index >= (int4)created.size() || seen[bl->index])
        indices_ok = false;
      else
        seen[bl->index] = true;
    }
    blocks << ']';
    for (int4 i = 0; i < seen.size(); ++i)
      if (!seen[i]) indices_ok = false;
    outedges << '[';
    bool first = true;
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      for (int4 s = 0; s < bl->outofthis.size(); ++s) {
        if (!first) outedges << ';';
        first = false;
        outedges << name(bl) << '!' << s << '>' << name(bl->outofthis[s].point)
                 << ':' << flagsText(bl->outofthis[s].label);
      }
    }
    outedges << ']';
    inedges << '[';
    first = true;
    for (int4 i = 0; i < created.size(); ++i) {
      FlowBlock *bl = created[i];
      for (int4 s = 0; s < bl->intothis.size(); ++s) {
        if (!first) inedges << ';';
        first = false;
        inedges << name(bl) << '@' << s << '<' << name(bl->intothis[s].point)
                << ':' << flagsText(bl->intothis[s].label);
      }
    }
    inedges << ']';
    std::cout << "case=" << caseName
              << "|preorder=" << pre.str()
              << "|roots=" << roots.str()
              << "|list=" << list.str()
              << "|blocks=" << blocks.str()
              << "|out=" << outedges.str()
              << "|in=" << inedges.str()
              << "|indices_ok=" << (indices_ok ? 1 : 0) << '\n';
  }
};

int main(void)
{
  std::cout << std::unitbuf;
  std::vector<std::string> spec_paths;
  startDecompilerLibrary(spec_paths);
  std::cout << "schema=1|fixture=BLOCK-INDEX-ASSIGN-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b\n";
  try {
    {
      Graph g;
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("empty_graph_early_return", preorder, rootlist);
    }
    {
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      (void)b0;
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("single_block", preorder, rootlist);
    }
    {
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b0, b2);
      g.edge(b1, b3);
      g.edge(b2, b3);
      g.edge(b0, b3);
      b1->index = 99;
      b1->visitcount = 5;
      b1->copymap = b0;
      b2->numdesc = 77;
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("single_entry_dag_corrupt_reset", preorder, rootlist);
    }
    {
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      g.edge(b0, b2);
      g.edge(b1, b2);
      g.edge(b2, b3);
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("multi_entry_rootlist_swap", preorder, rootlist);
    }
    {
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b1, b0);
      g.edge(b1, b2);
      g.edge(b2, b1);
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("no_root_assume_first_loop", preorder, rootlist);
    }
    {
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b2, b3);
      g.edge(b3, b2);
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("unreachable_extraroots_twopass", preorder, rootlist);
    }
    {
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      FlowBlock *b4 = g.makeBlock();
      g.edge(b0, b3);
      g.edge(b1, b0);
      g.edge(b2, b1);
      g.edge(b2, b2);
      g.edge(b4, b4);
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("stale_root_machinery", preorder, rootlist);
    }
    {
      Graph g;
      FlowBlock *b0 = g.makeBlock();
      FlowBlock *b1 = g.makeBlock();
      FlowBlock *b2 = g.makeBlock();
      FlowBlock *b3 = g.makeBlock();
      g.edge(b0, b1);
      g.edge(b0, b2);
      g.edge(b1, b3);
      g.edge(b2, b3);
      b0->setOutEdgeFlag(0, FlowBlock::f_irreducible | FlowBlock::f_back_edge);
      b1->setOutEdgeFlag(0, FlowBlock::f_goto_edge | FlowBlock::f_loop_exit_edge);
      std::vector<FlowBlock *> preorder;
      std::vector<FlowBlock *> rootlist;
      g.graph.findSpanningTree(preorder, rootlist);
      g.observe("irreducible_prelabel_wiped", preorder, rootlist);
    }
  }
  catch (const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
    return 1;
  }
  return 0;
}

File: tests/oracle/block_index_assign_1204.rs
// BLOCK-INDEX-ASSIGN-0001: Rugra comparand for the locked Ghidra 12.0.4
// BlockGraph::findSpanningTree (block.cc:1009-1136) oracle.  Mirrors
// tests/oracle/block_index_assign_1204.cc case for case: the same synthetic
// control-flow graphs are built through the production BlockGraph APIs
// (add_block/add_edge) and the complete post-call state is printed in the
// shared observation format.

use std::sync::{Arc, RwLock, Weak};

use rugra::address::Address;
use rugra::block::{edge_flags as ef, BlockBasic, BlockGraph, FlowBlock};

type BlockRef = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

fn flags_text(label: u32) -> String {
    let mut out = String::new();
    if label & ef::F_GOTO_EDGE != 0 { out.push('g'); }
    if label & ef::F_LOOP_EDGE != 0 { out.push('l'); }
    // Rugra's F_DEFAULTSWITCH_EDGE shares bit 7 with F_TREE_EDGE (pre-existing
    // edge_flags collision). Within this projection bit 7 can only be a tree
    // label: findSpanningTree wipes every edge label at the start of each pass
    // and never sets default-switch labels.
    if label & ef::F_DEFAULTSWITCH_EDGE != 0 && label & ef::F_TREE_EDGE == 0 { out.push('d'); }
    if label & ef::F_IRREDUCIBLE_EDGE != 0 { out.push('i'); }
    if label & ef::F_TREE_EDGE != 0 { out.push('t'); }
    if label & ef::F_FORWARD_EDGE != 0 { out.push('f'); }
    if label & ef::F_CROSS_EDGE != 0 { out.push('c'); }
    if label & ef::F_BACK_EDGE != 0 { out.push('b'); }
    if label & ef::F_LOOP_EXIT_EDGE != 0 { out.push('x'); }
    if out.is_empty() { out.push('-'); }
    out
}

struct Graph {
    graph: BlockGraph,
    created: Vec<BlockRef>,
    base: u64,
}

impl Graph {
    fn new(base: u64) -> Self {
        Graph { graph: BlockGraph::new(), created: Vec::new(), base }
    }

    fn make_block(&mut self) -> BlockRef {
        let bl: BlockRef = Arc::new(RwLock::new(BlockBasic::new(
            0,
            Address::new(self.base),
        )));
        self.graph.add_block(bl.clone());
        self.created.push(bl.clone());
        bl
    }

    fn edge(&mut self, from: &BlockRef, to: &BlockRef) {
        self.graph.add_edge(from.clone(), to.clone());
    }

    fn name(&self, bl: &BlockRef) -> String {
        for (i, cand) in self.created.iter().enumerate() {
            if Arc::ptr_eq(cand, bl) {
                return format!("b{}", i);
            }
        }
        "b?".to_string()
    }

    fn observe(&self, case_name: &str, preorder: &[BlockRef], rootlist: &[BlockRef]) {
        let pre: Vec<String> = preorder.iter().map(|b| self.name(b)).collect();
        let roots: Vec<String> = rootlist.iter().map(|b| self.name(b)).collect();
        let list: Vec<String> = self.graph.blocks.iter().map(|b| self.name(b)).collect();
        let mut indices_ok = true;
        let mut seen = vec![false; self.created.len()];
        let mut blocks = String::from("[");
        for (i, bl) in self.created.iter().enumerate() {
            if i != 0 { blocks.push(';'); }
            let g = bl.read().unwrap();
            blocks.push_str(&format!(
                "{}:index={},visit={},desc={},copy=",
                self.name(bl),
                g.get_index(),
                g.get_visit_count(),
                g.get_num_desc()
            ));
            let copy: Weak<RwLock<dyn FlowBlock + Send + Sync>> = match g.get_copy_map() {
                Some(w) => w,
                None => {
                    blocks.push_str("null");
                    continue;
                }
            };
            drop(g);
            match copy.upgrade() {
                Some(arc) if Arc::ptr_eq(&arc, bl) => blocks.push_str("self"),
                Some(arc) => blocks.push_str(&format!("other:{}", self.name(&arc))),
                None => blocks.push_str("null"),
            }
        }
        blocks.push(']');
        for bl in &self.created {
            let idx = bl.read().unwrap().get_index();
            if idx < 0 || idx as usize >= self.created.len() || seen[idx as usize] {
                indices_ok = false;
            } else {
                seen[idx as usize] = true;
            }
        }
        if seen.iter().any(|s| !s) { indices_ok = false; }
        let mut outedges = String::from("[");
        let mut first = true;
        for bl in &self.created {
            let g = bl.read().unwrap();
            for s in 0..g.size_out() {
                if let Some(e) = g.get_out(s) {
                    if !first { outedges.push(';'); }
                    first = false;
                    outedges.push_str(&format!(
                        "{}!{}>{}:{}",
                        self.name(bl),
                        s,
                        self.name(&e.point),
                        flags_text(e.flags)
                    ));
                }
            }
        }
        outedges.push(']');
        let mut inedges = String::from("[");
        first = true;
        for bl in &self.created {
            let g = bl.read().unwrap();
            for s in 0..g.size_in() {
                if let Some(e) = g.get_in(s) {
                    if !first { inedges.push(';'); }
                    first = false;
                    inedges.push_str(&format!(
                        "{}@{}<{}:{}",
                        self.name(bl),
                        s,
                        self.name(&e.point),
                        flags_text(e.flags)
                    ));
                }
            }
        }
        inedges.push(']');
        println!(
            "case={}|preorder=[{}]|roots=[{}]|list=[{}]|blocks={}|out={}|in={}|indices_ok={}",
            case_name,
            pre.join(","),
            roots.join(","),
            list.join(","),
            blocks,
            outedges,
            inedges,
            if indices_ok { 1 } else { 0 }
        );
    }
}

fn main() {
    println!("schema=1|fixture=BLOCK-INDEX-ASSIGN-0001|oracle=e40ed13014025f82488b1f8f7bca566894ac376b");

    // case=empty_graph_early_return
    {
        let mut g = Graph::new(0x4f00);
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("empty_graph_early_return", &preorder, &rootlist);
    }

    // case=single_block
    {
        let mut g = Graph::new(0x4f80);
        let _b0 = g.make_block();
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("single_block", &preorder, &rootlist);
    }

    // case=single_entry_dag_corrupt_reset
    {
        let mut g = Graph::new(0x5000);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        g.edge(&b0, &b3);
        b1.write().unwrap().set_index(99);
        b1.write().unwrap().set_visit_count(5);
        b1.write().unwrap().set_copy_map(Some(Arc::downgrade(&b0)));
        b2.write().unwrap().set_num_desc(77);
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("single_entry_dag_corrupt_reset", &preorder, &rootlist);
    }

    // case=multi_entry_rootlist_swap
    {
        let mut g = Graph::new(0x5100);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b2);
        g.edge(&b1, &b2);
        g.edge(&b2, &b3);
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("multi_entry_rootlist_swap", &preorder, &rootlist);
    }

    // case=no_root_assume_first_loop
    {
        let mut g = Graph::new(0x5200);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b1, &b0);
        g.edge(&b1, &b2);
        g.edge(&b2, &b1);
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("no_root_assume_first_loop", &preorder, &rootlist);
    }

    // case=unreachable_extraroots_twopass
    {
        let mut g = Graph::new(0x5300);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b2, &b3);
        g.edge(&b3, &b2);
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("unreachable_extraroots_twopass", &preorder, &rootlist);
    }

    // case=stale_root_machinery
    {
        let mut g = Graph::new(0x5400);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        let b4 = g.make_block();
        g.edge(&b0, &b3);
        g.edge(&b1, &b0);
        g.edge(&b2, &b1);
        g.edge(&b2, &b2);
        g.edge(&b4, &b4);
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("stale_root_machinery", &preorder, &rootlist);
    }

    // case=irreducible_prelabel_wiped
    {
        let mut g = Graph::new(0x5500);
        let b0 = g.make_block();
        let b1 = g.make_block();
        let b2 = g.make_block();
        let b3 = g.make_block();
        g.edge(&b0, &b1);
        g.edge(&b0, &b2);
        g.edge(&b1, &b3);
        g.edge(&b2, &b3);
        {
            // Pre-label both halves the way Ghidra's setOutEdgeFlag does.
            let (target, rev) = {
                let g0 = b0.read().unwrap();
                let e = g0.get_out(0).unwrap();
                (e.point.clone(), e.reverse_index)
            };
            b0.write().unwrap().set_out_edge_flag(0, ef::F_IRREDUCIBLE_EDGE | ef::F_BACK_EDGE);
            target.write().unwrap().set_in_edge_flag(rev as usize, ef::F_IRREDUCIBLE_EDGE | ef::F_BACK_EDGE);
        }
        {
            let (target, rev) = {
                let g1 = b1.read().unwrap();
                let e = g1.get_out(0).unwrap();
                (e.point.clone(), e.reverse_index)
            };
            b1.write().unwrap().set_out_edge_flag(0, ef::F_GOTO_EDGE | ef::F_LOOP_EXIT_EDGE);
            target.write().unwrap().set_in_edge_flag(rev as usize, ef::F_GOTO_EDGE | ef::F_LOOP_EXIT_EDGE);
        }
        let mut preorder = Vec::new();
        let mut rootlist = Vec::new();
        g.graph.find_spanning_tree(&mut preorder, &mut rootlist).expect("spanning tree");
        g.observe("irreducible_prelabel_wiped", &preorder, &rootlist);
    }
}
